transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java

```diff
@@ -84,17 +85,27 @@ public ChannelMetadata metadata() {
         return METADATA;
     }
 
+    final boolean shouldBreakReadReady(ChannelConfig config) {
+        return isInputShutdown0() && (inputClosedSeenErrorOnRead || !isAllowHalfClosure(config));
+    }
+
+    private static boolean isAllowHalfClosure(ChannelConfig config) {
+        return config instanceof DefaultSocketChannelConfig &&
+                ((DefaultSocketChannelConfig) config).isAllowHalfClosure();
```

Comment:
Replace `DefaultSocketChannelConfig` with `SocketChannelConfig`. After this its ready to be merged :)